src/practice4/task4/BlockingObjectPool.java

```diff
@@ -0,0 +1,69 @@
+package practice4.task4;
+
+import java.util.NoSuchElementException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+
+/**
+ * Pool that block when it has not any items or it full
+ */
+public class BlockingObjectPool {
+
+    private int capacity = 1;
+    private final BlockingQueue<Object> pool = new ArrayBlockingQueue<>(capacity);
```

Comment:
There is no sense to use your own synchronization logic if you have BlocingQueue. It is already synchronized. Use simple async storage